lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -3932,6 +3932,135 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
+
+  return classType.getElement(index).type;
+}
+
+void ObjectSubfieldOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
+  // no-op.
+}
+
+void ObjectSubfieldOp::print(OpAsmPrinter &p) {
+  auto input = getInput();
+  auto classType = input.getType();
+  p << ' ' << input << "[";
+  p.printKeywordOrString(classType.getElement(getIndex()).name);
+  p << "]";
+  SmallVector<StringRef, 1> elidedAttrs;
+  elidedAttrs.push_back("index");
+  p.printOptionalAttrDict((*this)->getAttrs(), elidedAttrs);
+  p << " : " << classType;
+}
+
+ParseResult ObjectSubfieldOp::parse(OpAsmParser &parser,
+                                    OperationState &result) {
+  auto *context = parser.getContext();
+
+  OpAsmParser::UnresolvedOperand input;
+  std::string fieldName;
+  Type inputType;
+  if (parser.parseOperand(input) || parser.parseLSquare() ||
+      parser.parseKeywordOrString(&fieldName) || parser.parseRSquare() ||
+      parser.parseOptionalAttrDict(result.attributes) || parser.parseColon() ||
+      parser.parseType(inputType))
+    return failure();
+
+  if (parser.resolveOperand(input, inputType, result.operands))
+    return failure();
+
+  auto classType = dyn_cast<ClassType>(inputType);
+  if (!classType)
+    return parser.emitError(parser.getNameLoc(),
+                            "input must be !firrtl.instance type, got ")
+           << inputType;
+
+  auto index = classType.getElementIndex(fieldName);
+  if (!index)
+    return parser.emitError(parser.getNameLoc(),
+                            "unknown field " + fieldName + " in instance type ")
+           << classType;
+  result.addAttribute("index",
+                      IntegerAttr::get(IntegerType::get(context, 32), *index));
+
+  SmallVector<Type> inferredReturnTypes;
+  if (failed(inferReturnTypes(context, result.location, result.operands,
+                              result.attributes.getDictionary(context),
+                              result.getRawProperties(), result.regions,
+                              inferredReturnTypes)))
+    return failure();
+  result.addTypes(inferredReturnTypes);
+
+  return success();
+}
+
+LogicalResult
+ObjectSubfieldOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  auto circuitOp = getOperation()->getParentOfType<CircuitOp>();
+  auto classType = getInput().getType();
+  auto className = classType.getNameAttr();
+
+  // verify that the class exists.
+  auto classOp = dyn_cast_or_null<ClassOp>(
+      symbolTable.lookupSymbolIn(circuitOp, className));
+  if (!classOp)
+    return emitOpError() << "target class '" << className.getValue()
+                         << "' not found";
+
+  // verify that the class type agrees with the class definition.
+  if (failed(classOp.verifyType(classType, [&]() { return emitOpError(); })))
+    return failure();
+
+  // verify that the field exists.
+  auto index = getIndex();
+  if (classType.getNumElements() < index)
```

Comment:
```suggestion
  if (classType.getNumElements() <= index)
```

Although this should be checked already by the normal verifier, probably the check below as well can be dropped?
Don't feel strongly so if it seems useful for some reason happy to keep it after minor fixups :+1: .

Reply:
I was doing all the verification in verify symbol uses, I've moved this check to the verifier, and also "the result type of this op agrees with the type of the field".